plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +118,15 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have a misbehaving upstream
+			// that is sending out-of-spec (>512 byte) UDP responses without an eDNS0
+			// OPT RR. Set the truncate bit and return the result - this will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			if proto == "udp" && strings.Contains(err.Error(), "overflow") {
```

Comment:
can we more accurately identify the error by type rather than string compares?

Reply:
I'd love to, but the [upstream message helpers that return the overflow errors](https://github.com/miekg/dns/blob/d83d31bedae5ce562b4686f7fbd70d4cba94424d/msg_helpers.go) aren't using any of [the exported defined errors from the package](https://github.com/miekg/dns/blob/master/msg.go#L51-L71).

Reply:
Is it worth pursuing an update in https://github.com/miekg/dns to type-out these errors? It requires waiting for a miekg/dns release and re-vendoring, so I'm assuming that would take longer.